modules/osm/intersection.js

```diff
@@ -410,18 +410,9 @@ export function osmIntersection(graph, startVertexId, maxDistance) {
                         if (t.id === way.id) {     // match VIA, TO
                             if (v.length === 1 && v[0].type === 'node' && v[0].id === entity.id) {
                                 matchesViaTo = true;    // match VIA node
-
-                            } else {                    // match all VIA ways
-                                var pathVias = [];
-                                for (k = 1; k < currPath.length; k++) {  // k = 1 skips FROM way
-                                    if (currPath[k][0] === 'w') pathVias.push(currPath[k]);
-                                }
-                                var restrictionVias = [];
-                                for (k = 0; k < v.length; k++) {
-                                    if (v[k].type === 'way') restrictionVias.push(v[k].id);
-                                }
-                                var diff = _difference(pathVias, restrictionVias);
-                                matchesViaTo = !diff.length;
+                            } else if (v.length === currPath.length / 2 - 1 &&
+                                _every(v, function(via, index) { return via.type === 'way' && currPath[(index+1)*2] === via.id; })) {
```

Comment:
I still don't completely understand this code...
It seems like the `_every` version of the code requires that the order of the multiple via ways must match the order in the path according to their index.  However the vias in `v` will not necessarily be ordered correctly.  They'll be ordered however they are stored in the Relation member array.

That's probably why I switched to gathering the restriction vias and the from vias into two arrays and using `_difference` to see if they have the same elements.

Reply:
> They'll be ordered however they are stored in the Relation member array.

oh. I thought the vias should be ordered in the osm relations… but I see that the [osm-wiki](https://wiki.openstreetmap.org/wiki/Relation:restriction) apparently doesn't specify the order… :thinking: could there be cases where in a really really complex intersection the order of traversed via-ways is also important?

Reply:
> could there be cases where in a really really complex intersection the order of traversed via-ways is also important?

It doesn't really have to be very complicated - but if there are 2 via ways, I think the algorithm needs to match the restriction regardless of what order the vias appear in the relation.
 
<img width="377" alt="screenshot 2018-04-10 09 44 48" src="https://user-images.githubusercontent.com/38784/38560757-5c3298cc-3ca4-11e8-9463-a08a6343f9c3.png">

Reply:
> does d2d464d look better?

I *think* so.. 🤔   The only situations I can think of that would break it would be invalid relations (like multiple via nodes, or mixed via node and way) and we dont need to worry about that..

Reply:
> oh. I thought the vias should be ordered in the osm relations… but I see that the osm-wiki apparently doesn't specify the order…

btw, I just thought more about this - another good reason for not having the via order matter: If you split a via way, the split code would need to intelligently reorder the vias (super complicated - see the issues we're having with PTv2 routes), rather than how it currently works of just making sure both split survivors end up as members.